internal/bits/minmax_amd64.go

```diff
@@ -0,0 +1,139 @@
+//go:build !purego
+
+package bits
+
+// The max-max algorithms combine looking for the min and max values in a single
+// pass over the data. While the behavior is the same as calling functions to
+// look for the min and max values independently, doing both operations at the
+// same time means that we only load the data from memory once. When working on
+// large arrays the algorithms are limited by memory bandwidth, computing both
+// the min and max together shrinks by half the amount of data read from memory.
+//
+// The following benchmarks results were highlighitng the benefits of combining
+// the min-max search, compared to calling the min and max functions separately:
+//
+// name                 old time/op    new time/op    delta
+// MinMaxInt64/10240KB     590µs ±15%     330µs ±10%  -44.01%  (p=0.000 n=10+10)
+//
+// name                 old speed      new speed      delta
+// MinMaxInt64/10240KB  17.9GB/s ±13%  31.8GB/s ±11%  +78.13%  (p=0.000 n=10+10)
+//
+// As expected, since the functions are memory-bound in those cases, and load
+// half as much data, we significant improvements. The gains are not 2x because
+// running more AVX-512 instructions in the tight loops causes more contention
+// on CPU ports.
+//
+//
+// As all things are always trade offs, using min/max functions independently
+// actually yields better throughput when the data resides in CPU caches:
```

Comment:
do you mean to say worse here? The benchmarks get worse in this case

Reply:
🤔 maybe I made the benchmarks hard to read

Reply:
In the benchmark results I've recorded the base is the min/max version, and it's compared to the combined min/max algorithm.

Reply:
The way that I read it, "old" is when the data is not residing in CPU caches, and "new" is when the data is residing in CPU caches. The "new" is both slower and has lower throughput in the benchmarks listed here.